decompiler/config/jak2/all-types.gc

```diff
@@ -24422,48 +24496,123 @@
 ;; water-h                        ;;
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
-#|
+(defenum water-flags
+  :bitfield #t
+  :type uint32
+  (active)
+  (can-wade)
+  (can-swim)
+  (swim-ground)
+  (can-ground)
+  (use-ocean)
+  (tar)
+  (mud)
+  (deadly)
+  (use-water-anim)
+  (no-grab-ground)
+  (dark-eco)
+  (lava)
+  (swamp)
+  (flow)
+  (under-water)
+  (head-under-water)
+  (bouncing)
+  (wading)
+  (swimming)
+  (touch-water)
+  (jump-out)
+  (break-surface)
+  (spawn-drip)
+  (part-splash)
+  (part-drip)
+  (part-rings)
+  (part-water)
+  (event)
+  (over-water)
+  (find-water)
+  (touch-water-good)
+  )
+
 (deftype water-info (structure)
-  ()
+  ((trans       vector      :offset 0)
+   (normal      vector      :offset 16)
+   (base-height meters      :offset 32)
+   (depth       meters      :offset 36)
+   (handle      uint64      :offset 40)
+   (flags       water-flags :offset 48)
+   (prim        basic       :offset 52)
+   (extra-flags uint32      :offset 56)
+   )
   :method-count-assert 9
   :size-assert         #x3c
   :flag-assert         #x90000003c
-  ;; Failed to read fields.
   )
-|#
 
-#|
 (deftype water-control (basic)
-  ()
+  ((flags             water-flags           :offset-assert 4)
+   (process           process-hidden        :offset-assert 8)
+   (joint-index       int32                 :offset-assert 12)
+   (top-y-offset      float                 :offset-assert 16)
+   (attack-id         uint32                :offset-assert 20)
+   (enter-water-time  time-frame            :offset-assert 24)
+   (wade-time         time-frame            :offset-assert 32)
+   (on-water-time     time-frame            :offset-assert 40)
+   (enter-swim-time   time-frame            :offset-assert 48)
+   (swim-time         time-frame            :offset-assert 56)
+   (base-height       meters                :offset-assert 64)
+   (wade-height       meters                :offset-assert 68)
+   (swim-height       meters                :offset-assert 72)
+   (surface-height    meters                :offset-assert 76)
+   (bottom-height     meters                :offset-assert 80)
+   (collide-height    meters                :offset-assert 84)
+   (height            meters                :offset-assert 88)
+   (height-offset     float       4         :offset-assert 92)
+   ;;(base-ocean-offset meters                :offset 92) ;; not sure what's going on with this float array...
+   ;;(real-ocean-offset meters       :score 1 :offset 92)
+   ;;(ocean-offset      meters                :offset 96)
+   ;;(bob-offset        meters                :offset 100)
```

Comment:
I think it should be fine to uncomment these? They would sometimes have a way to access the same data as either
```
(-> water-control height-offset 2)
```
or
```
(-> water-control bob-offset)
```
Another thing I like to do is put a `:score -1` on the `height-offset` field, so the decompiler will prefer using the more descriptive names when possible.

Reply:
(if this causes some other problem, then don't worry about it)

Reply:
Using `:score -1` on `height-offset` ended up working fine.